File: sim.cpp



#include <iostream>
#include <vector>
#include <stdlib.h>

/// g++ -o sim sim.cpp
/// ./sim


int main() {
    
    int n = 3 ; // number of doors
    int r = 1 ; // number of losing doors revealed by host, must be > 0 AND < n - 1
    int nn = 10000000 ; // total number of simulations
    
    int d_space[n] ;
    
    int door_prize ;
    int door_select ;
    
    int xndx_p, xndx_s ;
    
    int doors_can_reveal[ n - 1 ] ;
    int doors_revealed[ r ] ;
    int doors_can_switch_to[ n - r - 1 ] ;
    
    int doors_not_available_for_switch[ r + 1 ] ;
    
    //int xndx_cr ;
    
    int xcan_use ;
    
    int door_switch ;
    
    int i, j, jj, kk ;
    
    int bigSum_os = 0 ;
    int bigSum_sw = 0 ;
    
    double prop_sw, prop_os ;
    
    for(i=0; i<n; i++) {
        d_space[i] = i + 1 ;
    }
    
    for(i=0; i<nn; i++) {
        
        xndx_p = rand() % n ;
        xndx_s = rand() % n ;
        
        door_prize = d_space[ xndx_p ] ;
        door_select = d_space[ xndx_s ] ;
        
        kk = 0 ;
        for(j=0; j<n; j++) {
            if( (d_space[j] != door_prize) && (d_space[j] != door_select) ) {
                doors_can_reveal[kk] = d_space[ j ] ;
                kk = kk + 1 ;
            }
        }
        
        //// deterministic -- just first available
        for(j=0; j<r; j++) {
            doors_revealed[j] = doors_can_reveal[j] ;
        }
    
        
        doors_not_available_for_switch[0] = door_select ;
        for(jj=0; jj<r; jj++) {
            doors_not_available_for_switch[jj+1] = doors_revealed[jj] ;
        }
        
        for(j=0; j<n; j++) {
            xcan_use = 1 ;
            for(jj=0; jj<r+1; jj++) {
                if( d_space[ j ] == doors_not_available_for_switch[jj] ) {
                    xcan_use = 0 ;
                }
            }
            
            if(xcan_use == 1) { ////// deterministic -- first available
                door_switch = d_space[ j ] ;
                goto stop ;
            }
        }
        
    stop:
        
        if(door_prize == door_select) {
            bigSum_os = bigSum_os + 1 ;
        }
        
        if(door_prize == door_switch) {
            bigSum_sw = bigSum_sw + 1 ;
        }
          
        //std::cout << door_select << "\t" << door_prize << "\t" << doors_revealed[0] << "\t" << doors_can_switch_to[0] << "\n";
        
    }
    
    prop_os = static_cast<double>(bigSum_os) / nn ;
    prop_sw = static_cast<double>(bigSum_sw) / nn ;
    
   
    std::cout << "Total wins original selection: " << bigSum_os << "\n" << "Total wins switch: " << bigSum_sw << "\n" ;
    
    std::cout.precision(7) ;
    std::cout << "Proportion wins original selection: " << std::fixed << prop_os << "\n" << "Proportion wins switch: " << prop_sw << "\n" ;
    
    return 0 ;
}

File: sim.f90

! gfortran -o simf90 sim.f90

PROGRAM  MontyHallGameSim
	IMPLICIT  NONE

	INTEGER  :: n, r             ! number of doors, number of losing doors revealed by host
	INTEGER  :: nn=10000000      ! number of simulations
	LOGICAL  :: Cond_a, Cond_b   ! two logical conditions
   
	INTEGER   :: i, j, jj, kk
	INTEGER   :: door_prize, door_select, xndx_p, xndx_s, door_switch
   
	INTEGER, DIMENSION(:), ALLOCATABLE  :: d_space
   
	INTEGER, DIMENSION(:), ALLOCATABLE  :: doors_can_reveal
	INTEGER, DIMENSION(:), ALLOCATABLE  :: doors_revealed
	INTEGER, DIMENSION(:), ALLOCATABLE  :: doors_can_switch_to
	INTEGER, DIMENSION(:), ALLOCATABLE  :: doors_not_available_for_switch
  
	LOGICAL    :: xcan_use
	INTEGER    :: bigSum_os, bigSum_sw
	REAL       :: prop_sw, prop_os
	REAL       :: u
   
	!n = 3
	!r = 1

	WRITE(*,*) "enter number of doors . . ."
	READ(*,*)  n
	WRITE(*,*) "enter number of revealed losing doors . . ."
	READ(*,*)  r
   
	IF(r > n-2) THEN
		WRITE(*,*) "Number of revealed losing doors must be < n - 1 -- stopping"
		STOP
	END IF
   
	ALLOCATE( d_space(n) )   
    
	ALLOCATE( doors_can_reveal(n-1) ) 
	ALLOCATE( doors_revealed(r) ) 
	ALLOCATE( doors_can_switch_to(n-r-1) ) 
	ALLOCATE( doors_not_available_for_switch(r+1) ) 
   
	DO i = 1,n
		d_space(i) = i
	END DO
   
	bigSum_os = 0
	bigSum_sw = 0
  
	DO i = 1,nn
		CALL RANDOM_NUMBER(u)
		xndx_p = CEILING(u*n)
		CALL RANDOM_NUMBER(u)
		xndx_s = CEILING(u*n)
	
		door_prize  = d_space( xndx_p )   !since fortran is 1 indexed, can equivalently simply assign xndx_p
		door_select = d_space( xndx_s )   
	
		kk = 1
		DO j = 1,n
			IF( (d_space(j) /= door_prize) .AND. (d_space(j) /= door_select)  ) THEN
				doors_can_reveal(kk) = d_space(j) ;
				kk = kk + 1 ;
			END IF
		END DO
	
		DO j = 1,r
			doors_revealed(j) = doors_can_reveal(j)    ! deterministic
		END DO
	
		doors_not_available_for_switch(1) = door_select
		DO jj = 1,r
			doors_not_available_for_switch(jj+1) = doors_revealed(jj)
		END DO
	
	

		DO j = 1,n
			xcan_use = .true.
			DO jj = 1,r+1
				IF( d_space( j ) == doors_not_available_for_switch(jj) ) THEN
					xcan_use = .false.
				END IF
			END DO
           
			IF(xcan_use) THEN ! deterministic -- first available
				door_switch = d_space( j )
				GOTO 96
			END IF
		END DO
	
		!  goto here
		96 CONTINUE 
	   
	   
	   
		IF(door_prize == door_select) THEN
			bigSum_os = bigSum_os + 1
		END IF
        
		IF(door_prize == door_switch) THEN
			bigSum_sw = bigSum_sw + 1 ;
		END IF
	
	
	END DO 

	prop_os = REAL(bigSum_os) / nn ;
	prop_sw = REAL(bigSum_sw) / nn ;
   
	WRITE(*,*) "total number of simulations:                ",  nn
	WRITE(*,*) "number of wins with original selection:     ",  bigSum_os
	WRITE(*,*) "number of wins by switching:                ",  bigSum_sw
	WRITE(*,*) "proportion of wins with original selection: ",  prop_os
	WRITE(*,*) "proportion of wins by switching:            ",  prop_sw


END PROGRAM  MontyHallGameSim


